### ISSUE TITLE
Cannot build plugins on Mac with external htslib installation

### ISSUE THREAD

**Issue by dmckean** [samtools/bcftools#684]
Building bcftools that has been configured with `--with-htslib=/Users/dave/local/htslib/htslib-1.6` fails to build plugins, `gcc` seems to be unable to find the symbols linked from htslib. The bcftools executable is viable, but plugins cannot be built.  I believe the issue affects bcftools 1.5 as well. (output has been hard-wrapped with '\\')

```
gcc -rdynamic -pthread -L/Users/dave/local/htslib-1.6/lib  -o bcftools \
    main.o vcfindex.o tabix.o vcfstats.o vcfisec.o vcfmerge.o vcfquery.o vcffilter.o filter.o \
    vcfsom.o vcfnorm.o vcfgtcheck.o vcfview.o vcfannotate.o vcfroh.o vcfconcat.o vcfcall.o \
    mcall.o vcmp.o gvcf.o reheader.o convert.o vcfconvert.o tsv2vcf.o vcfcnv.o HMM.o \
    vcfplugin.o consensus.o ploidy.o bin.o hclust.o version.o regidx.o smpl_ilist.o csq.o \
    vcfbuf.o mpileup.o bam2bcf.o bam2bcf_indel.o bam_sample.o vcfsort.o ccall.o em.o \
    prob1.o kmin.o  polysomy.o peakfit.o -lhts -lm -lz -ldl  -lgsl -lcblas
clang: warning: argument unused during compilation: '-pthread'
gcc -g -O2 -I. -I/Users/dave/local/htslib-1.6/include   -c -o test/test-rbuf.o test/test-rbuf.c
gcc  -o test/test-rbuf test/test-rbuf.o -lz -ldl 
gcc -g -O2 -I. -I/Users/dave/local/htslib-1.6/include   -c -o test/test-regidx.o test/test-regidx.c
gcc -L/Users/dave/local/htslib-1.6/lib  -o test/test-regidx test/test-regidx.o regidx.o  -lpthread -lhts -lz -ldl 
gcc -bundle -bundle_loader bcftools -g -O2 -I. -I/Users/dave/local/htslib-1.6/include   \
     -o plugins/GTisec.so version.c plugins/GTisec.c 
Undefined symbols for architecture x86_64:
  "_bcf_get_format_values", referenced from:
      _process in GTisec-27bba6.o
  "_bcf_hdr_id2int", referenced from:
      _init in GTisec-27bba6.o
  "_bcf_unpack", referenced from:
      _process in GTisec-27bba6.o
  "_hts_version", referenced from:
      _version in version-ed3e88.o
      _init in GTisec-27bba6.o
ld: symbol(s) not found for architecture x86_64
clang: error: linker command failed with exit code 1 (use -v to see invocation)
make: *** [plugins/GTisec.so] Error 1
```

If I configure to use the built-in htslib, the issue goes away (see below). Of interest is that `bcftools` is explicitly linked to `libhts.a` . If I override `HTSLIB_LIB` to `htslib-1.6/libhts.a -lz -lm -lbz2 -llzma -lcurl` the issue is similarly resolved.

```
gcc -rdynamic -pthread   -o bcftools \
    main.o vcfindex.o tabix.o vcfstats.o vcfisec.o vcfmerge.o vcfquery.o vcffilter.o filter.o \
    vcfsom.o vcfnorm.o vcfgtcheck.o vcfview.o vcfannotate.o vcfroh.o vcfconcat.o vcfcall.o \
    mcall.o vcmp.o gvcf.o reheader.o convert.o vcfconvert.o tsv2vcf.o vcfcnv.o HMM.o \
    vcfplugin.o consensus.o ploidy.o bin.o hclust.o version.o regidx.o smpl_ilist.o csq.o \
    vcfbuf.o mpileup.o bam2bcf.o bam2bcf_indel.o bam_sample.o vcfsort.o ccall.o em.o \
    prob1.o kmin.o  htslib-1.6/libhts.a -lz -lm -lbz2 -llzma -lcurl  -lm -lz -ldl  
clang: warning: argument unused during compilation: '-pthread'
gcc -g -O2 -I. -Ihtslib-1.6   -c -o test/test-rbuf.o test/test-rbuf.c
gcc  -o test/test-rbuf test/test-rbuf.o -lz -ldl 
gcc -g -O2 -I. -Ihtslib-1.6   -c -o test/test-regidx.o test/test-regidx.c
gcc   -o test/test-regidx test/test-regidx.o regidx.o htslib-1.6/libhts.a htslib-1.6/libhts.a -lpthread htslib-1.6/libhts.a -lz -lm -lbz2 -llzma -lcurl  -lz -ldl 
gcc -bundle -bundle_loader bcftools -g -O2 -I. -Ihtslib-1.6    -o plugins/GTisec.so version.c plugins/GTisec.c 
```

I've tried adding `-Wl,-rpath=/.../htslib-1.6/lib` to the `LDFLAGS`, but that doesn't seem to fix the issue. Any suggestions?

**Comment by dmckean** [comment]
Update: also of note, I changed the external htslib dir to `/Users/dave/local/htslib-1.6/src/htslib-1.6`, where htslib was built, no issues.  Something different about linking to libraries in a built, source directory?

```
dave ~/local $ find htslib-1.6 -maxdepth 2
htslib-1.6
htslib-1.6/bin
htslib-1.6/bin/bgzip
htslib-1.6/bin/htsfile
htslib-1.6/bin/tabix
htslib-1.6/include
htslib-1.6/include/htslib
htslib-1.6/lib
htslib-1.6/lib/libhts.1.6.dylib
htslib-1.6/lib/libhts.2.dylib
htslib-1.6/lib/libhts.a
htslib-1.6/lib/libhts.dylib
htslib-1.6/lib/pkgconfig
htslib-1.6/libexec
htslib-1.6/libexec/htslib
htslib-1.6/share
htslib-1.6/share/man
htslib-1.6/src
htslib-1.6/src/htslib-1.6
htslib-1.6/src/htslib-1.6.tar.bz2

```

**Comment by dkj** [comment]
@mcshane keen for @daviesrob consultation

**Comment by daviesrob** [comment]
I've managed to reproduce this.  Building against an installed copy of HTSlib fails to build the plug-ins because the linker is missing an incantation to tell it what to do with undefined symbols.

For a quick fix, configure bcftools as before, and then try running:

```
make PLUGIN_FLAGS='-bundle -bundle_loader bcftools -Wl,-undefined,dynamic_lookup'
```

This should build successfully.  The more complete fix is to add `-Wl,-undefined,dynamic_lookup` to the line in bcftool's Makefile that sets `PLUGIN_FLAGS`.

If you take a look at how HTSlib builds its plug-ins on MacOS, you'll see that it uses the same linker option (and in fact this is how I discovered the solution).
